injected/src/features/message-bridge.js

```diff
@@ -10,56 +11,83 @@ import {
-        });
-        window.addEventListener(appendToken(ProxyRequest.NAME), (/** @type {CustomEvent<unknown>} */ e) => {
-            console.log(`[isolated:incoming] ${ProxyRequest.NAME}`, e.detail);
-            const request = ProxyRequest.create(e.detail);
-            if (!request) return;
-            this.proxyRequest(request, appendToken);
-        });
-        window.addEventListener(appendToken(SubscriptionRequest.NAME), (/** @type {CustomEvent<unknown>} */ e) => {
-            console.log(`[isolated:incoming] ${SubscriptionRequest.NAME}`, e.detail);
-            const subscription = SubscriptionRequest.create(e.detail);
-            if (!subscription) return;
-            this.proxySubscription(subscription, appendToken);
-        });
-        window.addEventListener(appendToken(SubscriptionUnsubscribe.NAME), (/** @type {CustomEvent<unknown>} */ e) => {
-            console.log(`[isolated:incoming] ${SubscriptionUnsubscribe.NAME}`, e.detail);
-            const unsubscribe = SubscriptionUnsubscribe.create(e.detail);
-            if (!unsubscribe) return;
-            this.removeSubscription(unsubscribe);
-        });
+
+        /**
+         * @param {{name: string; id: string} & Record<string, any>} incoming
+         */
+        const send = (incoming) => {
+            if (!args.messageSecret) return this.log('ignoring because args.messageSecret was absent');
+            const eventName = appendToken(incoming.name + '-' + incoming.id);
+            const event = new captured.CustomEvent(eventName, { detail: incoming });
+            captured.dispatchEvent(event);
+        };
+
+        /**
+         * @template T
+         * @param {{ create: (params: any) => T | null, NAME: string }} ClassType - A class with a `create` static method.
+         * @param {(instance: T) => void} callback - A callback that receives an instance of the class.
+         */
+        const accept = (ClassType, callback) => {
+            captured.addEventListener(appendToken(ClassType.NAME), (/** @type {CustomEvent<unknown>} */ e) => {
+                this.log(`${ClassType.NAME}`, e.detail);
+                const instance = ClassType.create(e.detail);
+                if (instance) {
+                    callback(instance);
+                } else {
+                    this.log('Failed to create an instance');
+                }
+            });
+        };
+
+        /**
+         * These are all the messages we accept from the page-world.
+         */
+        accept(InstallProxy, (install) => this.installProxyFor(install.featureName, args.messagingConfig));
+        accept(ProxyNotification, (notification) => this.proxyNotification(notification));
+        accept(ProxyRequest, (request) => this.proxyRequest(request, send));
+        accept(SubscriptionRequest, (subscription) => this.proxySubscription(subscription, send));
+        accept(SubscriptionUnsubscribe, (unsubscribe) => this.removeSubscription(unsubscribe.id));
```

Comment:
As far as I remember, `init()` is called asynchronously (not immediately on document start). I see that navigator interface is injected early on "privileged domains", which may create a race condition (page tries to use messaging before the proxy is ready).

Should we move this initialization to `load()`?

Reply:
or, maybe we should defer injecting the navigator interface until the proxy is ready?

Reply:
Yeah I'll verify all these things today

Reply:
Is it also true for the extension?
IIRC full config is not available at `load` stage, so feature might not be disable-able, so if you move, pls double check that